polkadot/node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -579,19 +580,26 @@ async fn determine_cores(
 	let cores = get_availability_cores(sender, relay_parent).await?;
 	let n_cores = cores.len();
 	let mut assigned_cores = Vec::new();
+	let maybe_claim_queue = fetch_claim_queue(sender, relay_parent).await?;
 
 	for (idx, core) in cores.iter().enumerate() {
-		let core_para_id = match core {
-			CoreState::Scheduled(scheduled) => Some(scheduled.para_id),
-			CoreState::Occupied(occupied) =>
-				if relay_parent_mode.is_enabled() {
-					// With async backing we don't care about the core state,
-					// it is only needed for figuring our validators group.
-					Some(occupied.candidate_descriptor.para_id)
-				} else {
-					None
-				},
-			CoreState::Free => None,
+		let core_para_id = match maybe_claim_queue {
+			Some(ref claim_queue) => {
+				//Runtime supports claim queue - use it
+				claim_queue.get_claim_for(CoreIndex::from(idx as u32), 0)
```

Comment:
I think this should be depth 1. Otherwise we have synchronous backing and we need to be real quick. I doubt this would work with 0 if the validaton takes 2s.

Reply:
I tested with 4 parachains sharing a core and candidates which take 1.8 seconds (any more and it times out on my laptop, which is not the best testing hardware). It all works

Reply:
this now leverages the full claim queue, resolving